qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -1217,16 +1217,63 @@ void ShaderViewer::ConfigureBookmarkMenu()
   QObject::connect(clearAction, &QAction::triggered, [this]() { ClearAllBookmarks(); });
   bookmarkMenu->addAction(clearAction);
 
-  QObject::connect(bookmarkMenu, &QMenu::aboutToShow, [this, nextAction, prevAction, clearAction]() {
-    const bool hasBookmarks = HasBookmarks();
-    nextAction->setEnabled(hasBookmarks);
-    prevAction->setEnabled(hasBookmarks);
-    clearAction->setEnabled(hasBookmarks);
-  });
+  QObject::connect(bookmarkMenu, &QMenu::aboutToShow,
+                   [this, bookmarkMenu, nextAction, prevAction, clearAction]() {
+                     UpdateBookmarkMenu(bookmarkMenu, nextAction, prevAction, clearAction);
+                   });
 
+  bookmarkMenu->addSeparator();
   ui->bookmark->setMenu(bookmarkMenu);
 }
 
+void ShaderViewer::UpdateBookmarkMenu(QMenu *menu, QAction *nextAction, QAction *prevAction,
+                                      QAction *clearAction)
+{
+  // setup permanent actions
+  const bool hasBookmarks = HasBookmarks();
+  nextAction->setEnabled(hasBookmarks);
+  prevAction->setEnabled(hasBookmarks);
+  clearAction->setEnabled(hasBookmarks);
+
+  // remove current bookmark list
+  QList<QAction *> actions = menu->actions();
+  for(auto itAction = actions.rbegin(); itAction != actions.rend() && !(*itAction)->isSeparator();
+      ++itAction)
+    menu->removeAction(*itAction);
+
+  // populate bookmark list
+  ScintillaEdit *cur = currentScintilla();
+  if(!cur)
+    return;
+
+  auto itBookmarks = m_Bookmarks.find(cur);
+  if(itBookmarks == m_Bookmarks.end())
+    return;
+
+  int numAddedBookmarks = 0;
+  for(auto &lineNumber : *itBookmarks)
+  {
+    QString textLine = QString::fromUtf8(cur->getLine(lineNumber)).simplified();
+    if(textLine.size() > BOOKMARK_MAX_MENU_ENTRY_LENGTH)
+    {
+      textLine.chop(textLine.size() - BOOKMARK_MAX_MENU_ENTRY_LENGTH);
+      textLine.append(lit("..."));
+    }
+    else if(textLine.isEmpty())
+    {
+      textLine = lit("(empty)");
+    }
+
+    QString name = QFormatStr("%1: %2").arg(lineNumber + 1).arg(textLine);
```

Comment:
I think it'd be better here to provide additional context and do the traditional `filename.ext:line` formulation rather than just the line number. That makes it easier for people to see that the menu is listing bookmarks in the current file - they could probably figure it out from the context but that's less direct.

Reply:
I tried implementing this but I feel like it makes the list of bookmarks a little harder to read:
![qrenderdoc_vRIZsG1a0k](https://github.com/baldurk/renderdoc/assets/16783712/bafd9b9f-f78d-49a4-8eb4-46de49a10730)

Here are two more ways of doing it.
![qrenderdoc_gXBA24mvkr](https://github.com/baldurk/renderdoc/assets/16783712/0b505d4f-d073-45a8-840e-8635ab550365)
![qrenderdoc_VffOiBYSUS](https://github.com/baldurk/renderdoc/assets/16783712/f6211a65-48ab-4786-8f1e-656a582715ad)

I could also specify the filename instead of "current file".

Thoughts?

Reply:
I think the readability problem there comes from the filename being a hash, not from including it in the entries. If the filename were human readable I think that would indeed be an improvement so I would like you to go with that route.

Maybe you could change the character limit to be say 10-20 chars longer but include the filename in that (still displaying the filename+line at minimum but displaying less text context).

Reply:
Maybe just add a couple more spaces between the line number and the text context would help?
Or some other separator character?

Reply:
We probably don't want to make the entries too wide unnecessarily, but `filename.hlsl:123   float declaration;` with a couple of extra spaces or `filename.hlsl:123 - float declaration;` would also be fine.

Reply:
Ok I ended up doing this:
![qrenderdoc_XURXQOr5Hs](https://github.com/baldurk/renderdoc/assets/16783712/09faafa3-e23d-4f43-8cdb-2af4f24c9294)